src/tlo/methods/wasting.py

```diff
@@ -1,41 +1,1947 @@
+                # schedule full recovery
+                self.sim.schedule_event(
+                    event=Wasting_FullRecovery_Event(module=self, person_id=person_id),
+                    date=outcome_date
+                )
+                # send for follow-up treatment for MAM
+                self.sim.modules['HealthSystem'].schedule_hsi_event(
+                    hsi_event=HSI_Wasting_SupplementaryFeedingProgramme_MAM(module=self, person_id=person_id),
+                    priority=0, topen=outcome_date
+                )
+
+            else:
+                outcome = self.rng.choice(['recovery_to_mam', 'death'],
+                                          p=[
+                                              1-self.parameters['prob_death_after_SAMcare'],
+                                              self.parameters['prob_death_after_SAMcare']
+                                          ])
+                if outcome == 'death':
+                    self.sim.schedule_event(
+                        event=Wasting_SevereAcuteMalnutritionDeath_Event(module=self, person_id=person_id),
+                        date=outcome_date
+                    )
+                    df.at[person_id, 'un_sam_death_date'] = outcome_date
+                else:  # recovery to MAM and send for treatment for MAM
+                    df.at[person_id, 'un_am_discharge_date'] = outcome_date
+                    self.sim.schedule_event(event=Wasting_RecoveryToMAM_Event(module=self, person_id=person_id),
+                                            date=outcome_date)
+                    self.sim.modules['HealthSystem'].schedule_hsi_event(
+                        hsi_event=HSI_Wasting_SupplementaryFeedingProgramme_MAM(module=self, person_id=person_id),
+                        priority=0, topen=outcome_date)
+
+    def cancel_future_event(self, person_id, event_type) -> None:
+        """
+        This function will add dates of recovery and/or progression events that need to be canceled.
+        :param person_id:
+        :param event_type: which event type to cancel
+        """
+
+        df = self.sim.population.props
+
+        event_tuples = [event_tuple for event_tuple in self.sim.find_events_for_person(person_id)
+                        if isinstance(event_tuple[1], event_type)]
+        if event_tuples:
+            dates = [event_tuple[0] for event_tuple in event_tuples]
+            event_type_map = {
+                Wasting_RecoveryToMAM_Event: 'un_recov_to_mam_to_cancel',
+                Wasting_FullRecovery_Event: 'un_full_recov_to_cancel',
+                Wasting_ProgressionToSevere_Event: 'un_progression_to_cancel'
+            }
+
+            # add dates of natural history events to be cancelled due to tx, the health outcome will be driven by the tx
+            for date in dates:
+                df.at[person_id, event_type_map[event_type]].append(date)
+
+        # else: event_type not scheduled, hence no need to cancel any
```

Comment:
this is not the way to cancel events. The look of of "find_events_for_person" is ONLY for testing/debugging as it's extremely slow.

What you need to do is -- in the event itself -- check whether it should still run. And if not, then do not do anything.

This check can either be:...
....  a flag stored on the instance of the event itself: this requires you to store pointers to the event.
..... through looking at the population dataframe: e.g. store the date you want the event to run, and if you don't want it to run, change it do pd.NaT... then when the event run, only let it do something if the today's date is the date stored in the datafrmae